### ISSUE TITLE
Kotlin code generator ignoring isKotlinNotNullRecordAttributes = true or I misunderstand this feature.

### ISSUE THREAD

**Issue by apatrida** [jOOQ/jOOQ#14948]
### Expected behavior

When in Gradle code generator configuration for KotlinGenerator, these values are set: `isKotlinNotNullRecordAttributes = true` the non nullable columns in a table will still be nullable in Kotlin records instead of the expected not null.

### Actual behavior

The non-nullable columns are still nullable.

### Steps to reproduce the problem

in Gradle `isRecords = true` and `isKotlinNotNullRecordAttributes = true`

a MySQL 8 table with primary key not nullable auto_increment integer.

For the table field:

```
val ID: TableField<SomethingRecord, Int?> = createField(DSL.name("id"), SQLDataType.INTEGER.nullable(false).identity(true), this, "")
```

this generates record field

```
open var id: Int?
        set(value): Unit = set(0, value)
        get(): Int? = get(0) as Int?
```

### jOOQ Version

3.18.3

### Database product and version

MySQL 8.0.26 RDS Aurora

### Java Version

19

### OS Version

MAC

### JDBC driver name and version (include name if unofficial driver)

com.mysql:mysql-connector-j:8.0.32

**Comment by apatrida** [comment]
This project is using OSS edition

**Comment by lukaseder** [comment]
Thanks for your report. This is a duplicate of:

- https://github.com/jOOQ/jOOQ/issues/14786
- https://github.com/jOOQ/jOOQ/issues/14787

**Comment by lukaseder** [comment]
I guess the ideas that you once presented with different object lifecycles for different DB interactions (`INSERT`, `UPDATE`, etc.) would be the only way to solve this in an intuitive way, given that this question re-appears so frequently.

If one record type has to fit all purposes, then I'm afraid, the presence of `DEFAULT` or some identity expression means nullable in client code.

**Comment by apatrida** [comment]
Yes, I see.  Back to my own code generator then for some playground work.

**Comment by apatrida** [comment]
Note @lukaseder that in my prototype I use 3-state instead of null to represent default value, because NULL does not mean "use default" in SQL.  If you pass a NULL to SQL (assuming nullable field with default) it doesn't use the default.

There are two ways in Kotlin to easily do 3 state:  

* First, use a delegate so you can observe if the value is ever set or not, it not being set uses default.  If it is set, it is not the default.  You do not need it to be nullable to detect "DEFAULT" as the change tracking does this.
* Second, use a tri-type backing field for a public property that changes from defaulted to one of the values (literal value or SQL expression)

Java obviously has ways with getter/setters to do similar.  You can even violate the JavaBean spec and have a setter for literal value and a setter for SQL expression.

The NULL signals default, starts this whole problem on insert.

**Comment by lukaseder** [comment]
> Note @lukaseder that in my prototype I use 3-state instead of null to represent default value, because NULL does not mean "use default" in SQL. If you pass a NULL to SQL (assuming nullable field with default) it doesn't use the default.

Just like jOOQ's `UpdatableRecord` (calling the setter with `null` != omitting calling the setter). jOOQ has implemented this correctly since version 1.0.

But this topic here isn't about jOOQ's `UpdatableRecord` internals,  but how it represents the types in kotlin through its API...
